File: app/src/main/java/com/productivity/habits/ui/daily/SlotHabitControls.kt
package com.productivity.habits.ui.daily

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.productivity.habits.data.local.entity.HabitEntity
import com.productivity.habits.data.local.entity.HabitLogEntity
import com.productivity.habits.domain.engine.SubdaySlotEngine
import com.productivity.habits.ui.common.HapticsHelper

@Composable
fun SlotHabitControls(
    habit: HabitEntity,
    logsForDate: List<HabitLogEntity>,
    accentColor: Color,
    onToggleSlot: (Int) -> Unit,
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current
    val haptic = LocalHapticFeedback.current

    val slots = SubdaySlotEngine.generateSlots(habit, logsForDate)
    val completedCount = slots.count { it.completed }
    val totalSlots = slots.size

    Row(
        modifier = modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        slots.forEach { slot ->
            val isDone = slot.completed
            val backgroundColor = if (isDone) accentColor else MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f)
            val contentColor = if (isDone) Color.White else MaterialTheme.colorScheme.onSurfaceVariant
            val borderStroke = if (isDone) null else BorderStroke(1.dp, MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.7f))

            Surface(
                modifier = Modifier
                    .clip(RoundedCornerShape(12.dp))
                    .clickable {
                        val willCompleteAll = !isDone && (completedCount + 1 >= totalSlots)
                        if (willCompleteAll) {
                            HapticsHelper.performHeavyConfirmationHaptic(context, haptic)
                        } else {
                            HapticsHelper.performLightHaptic(haptic)
                        }
                        onToggleSlot(slot.index)
                    },
                shape = RoundedCornerShape(12.dp),
                color = backgroundColor,
                border = borderStroke
            ) {
                Row(
                    modifier = Modifier.padding(horizontal = 10.dp, vertical = 6.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    if (isDone) {
                        Icon(
                            imageVector = Icons.Default.Check,
                            contentDescription = "Completed",
                            tint = Color.White,
                            modifier = Modifier.size(14.dp)
                        )
                        Spacer(modifier = Modifier.width(4.dp))
                    }
                    Text(
                        text = slot.timeLabel,
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = if (isDone) FontWeight.Bold else FontWeight.Medium,
                        color = contentColor
                    )
                }
            }
        }
    }
}

File: lib/ui/daily/widgets/slot_habit_controls.dart
import 'package:flutter/material.dart';
import '../../../domain/engines/subday_slot_engine.dart';
import '../../../domain/models/habit.dart';
import '../../../domain/models/habit_log.dart';
import '../../common/haptics_helper.dart';

class SlotHabitControls extends StatelessWidget {
  final Habit habit;
  final List<HabitLog> logsForDate;
  final Color accentColor;
  final ValueChanged<int> onToggleSlot;

  const SlotHabitControls({
    super.key,
    required this.habit,
    required this.logsForDate,
    required this.accentColor,
    required this.onToggleSlot,
  });

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final slots = SubdaySlotEngine.generateSlots(habit, logsForDate);
    final completedCount = slots.where((s) => s.completed).length;
    final totalSlots = slots.length;

    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Row(
        children: slots.map((slot) {
          final isDone = slot.completed;
          final backgroundColor = isDone
              ? accentColor
              : theme.colorScheme.surfaceContainerHighest
                  .withValues(alpha: 0.5);
          final contentColor =
              isDone ? Colors.white : theme.colorScheme.onSurfaceVariant;
          final borderSide = isDone
              ? BorderSide.none
              : BorderSide(
                  color: theme.colorScheme.outlineVariant
                      .withValues(alpha: 0.7),
                );

          return Padding(
            padding: const EdgeInsets.only(right: 8),
            child: Material(
              color: backgroundColor,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(12),
                side: borderSide,
              ),
              child: InkWell(
                borderRadius: BorderRadius.circular(12),
                onTap: () {
                  final willCompleteAll =
                      !isDone && (completedCount + 1 >= totalSlots);
                  if (willCompleteAll) {
                    HapticsHelper.performHeavyConfirmationHaptic();
                  } else {
                    HapticsHelper.performLightHaptic();
                  }
                  onToggleSlot(slot.index);
                },
                child: Padding(
                  padding: const EdgeInsets.symmetric(
                    horizontal: 10,
                    vertical: 6,
                  ),
                  child: Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      if (isDone) ...[
                        const Icon(
                          Icons.check,
                          color: Colors.white,
                          size: 14,
                        ),
                        const SizedBox(width: 4),
                      ],
                      Text(
                        slot.timeLabel,
                        style: theme.textTheme.labelMedium?.copyWith(
                          fontWeight:
                              isDone ? FontWeight.bold : FontWeight.w500,
                          color: contentColor,
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
          );
        }).toList(),
      ),
    );
  }
}
